examples/java-matter-controller/java/src/com/matter/controller/commands/common/Command.java

```diff
@@ -0,0 +1,292 @@
+    Argument arg = new Argument();
+    arg.type = ArgumentType.String;
+    arg.name = name;
+    arg.value = (Object) out;
+    arg.desc = Optional.ofNullable(desc);
+
+    mArgs.add(arg);
+    return mArgs.size();
+  }
+
+  /**
+   * @brief Add an optional long integer command argument
+   * @param name The name that will be displayed in the command help
+   * @param min The minimum value of the argv value
+   * @param max The minimum value of the argv value
+   * @param out A pointer to an Optional where the argv value will be stored, the argv value could
+   *     be empty
+   * @param desc The description of the argument that will be displayed in the command help
+   * @return The number of arguments currently added to the command
+   */
+  public final int addArgument(String name, long min, long max, Optional<Long> out, String desc) {
+    Argument arg = new Argument();
+    arg.type = ArgumentType.Number_int64;
+    arg.name = name;
+    arg.value = (Object) out;
+    arg.min = min;
+    arg.max = max;
+    arg.flags = Argument.kOptional;
+    arg.desc = Optional.ofNullable(desc);
+
+    mArgs.add(arg);
+    return mArgs.size();
+  }
+
+  /**
+   * @brief Add an optional short integer command argument
+   * @param name The name that will be displayed in the command help
+   * @param min The minimum value of the argv value
+   * @param max The minimum value of the argv value
+   * @param out A pointer to an Optional where the argv value will be stored, the argv value could
+   *     be empty
+   * @param desc The description of the argument that will be displayed in the command help
+   * @return The number of arguments currently added to the command
+   */
+  public final int addArgument(
+      String name, short min, short max, Optional<Short> out, String desc) {
+    Argument arg = new Argument();
+    arg.type = ArgumentType.Number_int16;
+    arg.name = name;
+    arg.value = (Object) out;
+    arg.min = min;
+    arg.max = max;
+    arg.flags = Argument.kOptional;
+    arg.desc = Optional.ofNullable(desc);
+
+    mArgs.add(arg);
+    return mArgs.size();
+  }
+
+  /**
+   * @brief Add an optional boolean command argument
+   * @param name The name that will be displayed in the command help
+   * @param out A pointer to an Optional where the argv value will be stored, the argv value could
+   *     be empty
+   * @param desc The description of the argument that will be displayed in the command help
+   * @return The number of arguments currently added to the command
+   */
+  public final int addArgument(String name, Optional<Boolean> out, String desc) {
+    Argument arg = new Argument();
+    arg.type = ArgumentType.Bool;
+    arg.name = name;
+    arg.value = (Object) out;
+    arg.flags = Argument.kOptional;
+    arg.desc = Optional.ofNullable(desc);
+
+    mArgs.add(arg);
+    return mArgs.size();
+  }
+
+  public final void initArguments(int argc, String[] args) throws IllegalArgumentException {
```

Comment:
Can you document this method? Is `argc` supposed to be the "required argument count"?

Reply:
Also, could this not be part of the constructor? This way you can enforce the arguments are initialized exactly once.

Reply:
Do you mean add to the constructor? The arguments are constructed and added to mArgs during command registration,  but we only know how to initialize them when we receive the command-line arguments for a command.

Reply:
I see, so `Argument` represents both a specification for argument construction, as well as consuming them. 

Not blocking, but IMO I would probably separate the two into something like:

`ArgumentSpec` and `Argument` (where `ArgumentSpec` could take in the actual command-line values and return an `Argument` after validating the inputs). This way `Argument` is never in a potentially invalid state while it's being initialized.

Reply:
Yeah I think I don't quite have a full picture of the flow end to end, but conceptually I think of it as:

You have some command definition (OnCommand) that can be run with some particular types of arguments (ArgumentSpecs). When the command is actually invoked, you would run it with some specific argument values (Argument). Running the command with those arguments would validate that they are the expected type and range, and execute if the validation passes, otherwise they would throw an illegal argument exception.

But yes, totally fine to defer this.

Reply:
It looks we need maintain two dynamic lists within the Command class, the Command has a list of ArgumentSpecs during registration, when the command is actually invoked, we initialize the ArgumentSpecs list and generate a valid Argument for each of them and add it to another Argument list, then we run the command against  Argument list instead of ArgumentSpecs list.

The current approach is we only main one list of Argument in class Command, the fill the specification part during registration, when the command is actually invoked, we initialize the Argument list in space and throw an illegal argument exception if hit an error, then run the command if the argument list has been successfully initialized.  

The latter approach is adopted by chip-tool.